ansible/collections/ansible_collections/nhsd/apigee/plugins/module_utils/apigee_action.py

```diff
@@ -0,0 +1,84 @@
+import json
+
+import ansible
+import pydantic
+import deepdiff
+import requests
+
+
+class ApigeeAction(ansible.plugins.action.ActionBase):
+
+    def validate_args(self, Validator):
+        """
+        Returns two-length tuple of validated_args and errors (dict).
+        """
+        try:
+            args = Validator(**self._task.args)
+        except pydantic.ValidationError as e:
+            return None, {
+                "failed": True,
+                "message": f"Validation of ansible module arguments failed",
+                "args": self._task.args,
+                "validation error": json.loads(e.json())
+            }
+
+        return args, {}
+
+    def delta(self, before, after, keys_to_ignore=None):
+        if not keys_to_ignore:
+            keys_to_ignore = []
+        return json.loads(
+            deepdiff.DeepDiff(
+                before,
+                after,
+                ignore_order=True,
+                ignore_type_in_groups=[(ansible.utils.unsafe_proxy.AnsibleUnsafeText, str)],
+                exclude_paths=[f"root['{key}']" for key in keys_to_ignore],
+            ).to_json()
+        )
+
+    def request(self, method, url, access_token, json=None, headers=None, status_code=None):
+
+        if not status_code:
+            status_code = [200]
+
+        if not headers:
+            headers = {}
+
+        headers["authorization"] = f"Bearer {access_token}"
+
+        response = requests.request(method, url,
+                                    json=json,
+                                    headers=headers)
+
+        response_dict = {
+            "response": {"status_code": response.status_code, "reason": response.reason},
+            "allowed_status_codes": status_code,
+        }
+
+        if response.status_code not in status_code:
+            response_dict["failed"] = True
+
+        if len(response.content.decode()) == 0:
```

Comment:
nitpick but this is being decoded at least twice in each branch, would better to save to var